cupy/cuda/cufft.pyx

```diff
@@ -168,6 +434,171 @@ class Plan1d(object):
+                # CUFFT_XT_FORMAT_INPLACE_SHUFFLED. For reuse we must correct
+                # it, otherwise in the next run we would encounter
+                # CUFFT_INVALID_TYPE!
+                ptr = self.xtArr
+                xtArr = <XtArray*>ptr
+                if self.batch == 1:
+                    if isinstance(a, cupy.ndarray):
+                        fmt = CUFFT_XT_FORMAT_INPLACE
+                    else:  # from numpy
+                        fmt = CUFFT_XT_FORMAT_1D_INPUT_SHUFFLED
+                    xtArr.subFormat = fmt
+        elif isinstance(a, list):
+            # TODO(leofang): For users running Plan1d.fft() (bypassing all
+            # checks in cupy.fft.fft), they are allowed to send in a list of
+            # ndarrays, each of which is on a different GPU. Then, no data
+            # copy is needed, just replace the pointers in the descriptor.
+            raise NotImplementedError('User-managed buffer area is not yet '
+                                      'supported.')
+        else:
+            raise ValueError('Impossible to reach.')
+
+    def _multi_gpu_memcpy(self, a, str action):
+        cdef Handle plan = self.plan
+        cdef list xtArr_buffer, share
+        cdef int start, nGPUs, i, count, result
+        cdef XtArray* arr
+        cdef intptr_t ptr, ptr2
+        cdef size_t size
+
+        if isinstance(a, cupy.ndarray) or isinstance(a, numpy.ndarray):
+            start = 0
+            b = a.ravel()
+            assert b.flags['OWNDATA'] is False
+            assert self.xtArr_buffer is not None
+            ptr = self.xtArr
+            arr = <XtArray*>ptr
+            xtArr_buffer = self.xtArr_buffer
+            nGPUs = len(self.gpus)
+            share = self.batch_share
+
+            if action == 'scatter':
+                if isinstance(a, cupy.ndarray):
+                    for i in range(nGPUs):
+                        count = int(share[i] * self.nx)
+                        size = count * b.dtype.itemsize
+                        curr_stream = self.streams[i]
+                        if i == 0:
+                            # When we come here, another stream could still be
+                            # copying data for us, so we wait patiently...
+                            outer_stream = stream_module.get_current_stream()
+                            outer_stream.synchronize()
+                        xtArr_buffer[i].copy_from_device_async(
+                            b[start:start+count].data, size, curr_stream)
+                        if i != 0:
+                            prev_event = self.events[i-1]
+                            curr_stream.wait_event(prev_event)
+                        curr_event = self.events[i]
+                        curr_event.record(curr_stream)
+                        if i == nGPUs - 1:
+                            curr_event.synchronize()
+                        start += count
+                    assert start == b.size
```

Comment:
Given that the multi-GPU FFT itself is blastingly fast, I think this naive "scattering" scheme could be improved, such as using a tree (so that the intermediate nodes can do both copy-in and copy-out at the same time). Too bad NCCL does not have a collective scatter call...

Reply:
When the main bottleneck is in data transfer rather than in communication overhead, a tree will be more costly than a plain loop. So for the purpose here we can't do better. Marking this as resolved.